plugins/uicommon.h

```diff
@@ -198,13 +198,14 @@ static inline char get_string_input(const std::set<df::interface_key> *input)
 
 static inline df::building_stockpilest *get_selected_stockpile()
 {
-    if (!Gui::dwarfmode_hotkey(Core::getTopViewscreen()) ||
+    return nullptr;
```

Comment:
Was this not working in v50? As far as I can tell, it should have been compiling. If there's an ability to select a building in DF (which I'm pretty sure there is), we should be returning the selected building as before.

Reply:
Would it make sense to migrate this function to `Gui:getAnyStockpile()`/`Gui:getSelectedStockpile()`?

Reply:
```
df::building_stockpilest* Gui::getAnyStockpile(df::viewscreen* top) {
    if (Gui::getFocusString(Core::getTopViewscreen())[0] == "dwarfmode/Some/Stockpile") {
        return game->main_interface.stockpile.cur_bld;
    }

    return NULL;
}

df::building_stockpilest* Gui::getSelectedStockpile(color_ostream& out, bool quiet) {
    df::building_stockpilest* stockpile = getAnyStockpile(Core::getTopViewscreen());

    if (!stockpile && !quiet)
        out.printerr("No stockpile is selected in the UI.\n");

    return stockpile;
}
```

?

Reply:
I'm actually surprised that it's not already returned by getSelectedBuilding() -- stockpiles are just buildings. If callers care if it's a stockpile, they can check themselves.

Reply:
if it's covered by getSelectedBuilding(), I'd say we don't even need it. Adding a getSelectedFoo() call comes with quite a bit of overhead since it then needs to be passed through a lot of other tools that implement the interface

Reply:
Makes sense! I think the [current implementation](https://github.com/DFHack/dfhack/blob/develop/library/modules/Gui.cpp#L1279-L1281) of `getSelectedBuilding` won't cover it because it doesn't use view sheets, but that should be easy enough to fix.

Reply:
My eventual plan is to move useful stuff out of uicommon and into core, then delete that file. Option 2 above is good.

Reply:
I don't think we need `getAnyStockpile`, just `getSelectedStockpile`. `getAnyBuilding`/`getSelectedBuilding` already return the selected building, and the only purpose of `get_selected_stockpile()` is to filter out any selected buildings that aren't stockpiles. It's just a thin wrapper around `getSelectedBuilding()`, even though it isn't actually implemented as such.

It is probably also true that `get_selected_stockpile()` could be replaced with `virtual_cast<df::building_stockpilest>(getSelectedBuilding())`. I think that's true in all cases...

Reply:
So far I've implemented option 2 above. I really feel like this approach makes the most sense since we retrieve the data about which stockpile is active from a different place in `game` than where we retrieve the building data.

If I make the suggested change, `getSelectedBuilding` would need to be updated to check another spot in `game` to return the currently selected stockpile, before or after checking for a building (`main_interface.view_sheets.active_id` vs `game->main_interface.stockpile.cur_bld`), right? Is that a better tradeoff than implementing the methods as is? Apologies in advance if I'm misunderstanding 😅.

Either way, happy to do whichever approach is considered best!